### PR TITLE
[FIX] ensure that multi maskers can clean with confounds and high variance confounds

### LINKED ISSUE #5313: multi maskers do not seem to support high_variance_confounds
multi mmaskers do not seem to support high_variance_confounds

_Originally posted by @Remi-Gau in https://github.com/nilearn/nilearn/pull/5291#discussion_r2025422685_

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #5313 

Changes proposed in this pull request:

-
-

### CODE REVIEW

File: nilearn/glm/tests/test_first_level.py [hunk 1]
```diff
@@ -2383,11 +2383,7 @@ def test_flm_compute_contrast_with_surface_data(surface_glm_data):
 def test_flm_get_element_wise_model_attribute_with_surface_data(
     surface_glm_data,
 ):
-    """Smoke test 'voxel wise' attribute with surface data.
-
-    TODO: rename the private function _get_element_wise_model_attribute
-    to work for both voxel and vertex
-    """
+    """Smoke test 'voxel wise' attribute with surface data."""
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:2386]: unrelated: just removing a obsolete TODO

File: nilearn/maskers/multi_nifti_masker.py [hunk 2]
```diff
@@ -542,12 +543,75 @@ def transform(self, imgs, confounds=None, sample_mask=None):
 
         """
         check_is_fitted(self)
-        if not hasattr(imgs, "__iter__") or isinstance(imgs, str):
-            return self.transform_single_imgs(imgs)
```
Comment (Remi-Gau) [inline nilearn/maskers/multi_nifti_masker.py:546]: for all multimaskers confounds and sample_mask were not applied if a nifti-like object was passed

when a list of nifti-like object was passed, they could not handle properly the expected diversity on inputs array, dataframe, csv file...

File: nilearn/_utils/estimator_checks.py [hunk 3]
```diff
@@ -1499,16 +1557,84 @@ def check_multi_masker_transformer_sample_mask(estimator):
...
+                [input_img, input_img], confounds=confounds
+            )
+
+            estimator = clone(estimator)
+            estimator.high_variance_confounds = True
+            signal_c_hvc = estimator.fit_transform(
+                [input_img, input_img], confounds=confounds
+            )
+
+            for s1, s2 in zip(signal_c, signal_c_hvc):
```
Comment (bthirion) [inline nilearn/_utils/estimator_checks.py:1634]: why do we need to do a zip here ?

**Review by bthirion — COMMENTED** [review]
LGTM besides one small detail. Thx !

**Review by hndgzkn — APPROVED** [review]
Some comments, otherwise seems Ok to me.

File: nilearn/_utils/estimator_checks.py (hunk 3 above)
Comment (Remi-Gau) [reply nilearn/_utils/estimator_checks.py:1634]: because the transform() of multimaskers return list of arrays so we can to compare them in a pairwise fashion

File: nilearn/_utils/estimator_checks.py [hunk 4]
```diff
@@ -1273,22 +1312,20 @@ def check_nifti_masker_fit_transform(estimator):
...
-    assert isinstance(signal, np.ndarray)
-    assert signal.shape == (1, estimator.n_elements_)
+        assert isinstance(signal, np.ndarray)
+        assert signal.shape[1] == estimator.n_elements_
 
-    estimator.fit_transform(_img_3d_rand())
+    signal = estimator.fit_transform(_img_3d_rand())
 
     assert isinstance(signal, np.ndarray)
     assert signal.shape == (1, estimator.n_elements_)
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/_utils/estimator_checks.py:1331]: am gonna leave as is because it needs to modified in https://github.com/nilearn/nilearn/pull/5381 as well

so let's try to avoid some extra churn

File: nilearn/maskers/multi_nifti_masker.py [hunk 5]
```diff
@@ -542,12 +543,75 @@ def transform(self, imgs, confounds=None, sample_mask=None):
...
+            )
+        if isinstance(imgs, NiimgLike):
+            confounds = (
+                confounds[0] if isinstance(confounds, list) else confounds
+            )
+            sample_mask = (
+                sample_mask[0]
+                if isinstance(sample_mask, list)
+                else sample_mask
+            )
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/maskers/multi_nifti_masker.py:564]: yeah but we do not quite have a parent class just for multimaskers and I would prefer to avoid a putting in BaseMasker a method that is not used by all children

File: nilearn/maskers/multi_nifti_masker.py [hunk 6]
```diff
@@ -517,20 +514,24 @@ def transform(self, imgs, confounds=None, sample_mask=None):
 
         Parameters
         ----------
-        imgs : :obj:`list` of Niimg-like objects
+        imgs : Niimg-like object, or a :obj:`list` of Niimg-like objects
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/maskers/multi_nifti_masker.py:516]: yeah I know: that kinda of bugs me too

what I was trying to do here is to make sure confounds are used properly without changing the API

multimaskers are ideally meant to work with list of images, but the current behavior allows a single image to be passed.

I considered that allowing something else that lists for confounds and mask would in make the multimaskers too much like their non multi counterpart
